File: backend/src/utils/similarity.ts
// TF-IDF based similarity calculation for product recommendations
// Implements cosine similarity without external dependencies

import { Product } from '@prisma/client';
import { DocumentVector } from '../types/index.js';

/**
 * Tokenize text into words (lowercase, no punctuation)
 */
function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .split(/\s+/)
    .filter(token => token.length > 0);
}

/**
 * Calculate term frequency for a document
 */
function calculateTF(tokens: string[]): Map<string, number> {
  const tf = new Map<string, number>();
  const totalTokens = tokens.length;

  for (const token of tokens) {
    tf.set(token, (tf.get(token) || 0) + 1);
  }

  // Normalize by document length
  for (const [term, count] of tf.entries()) {
    tf.set(term, count / totalTokens);
  }

  return tf;
}

/**
 * Calculate inverse document frequency
 */
function calculateIDF(documents: string[][]): Map<string, number> {
  const idf = new Map<string, number>();
  const totalDocs = documents.length;

  // Count documents containing each term
  const docFrequency = new Map<string, number>();
  for (const doc of documents) {
    const uniqueTerms = new Set(doc);
    for (const term of uniqueTerms) {
      docFrequency.set(term, (docFrequency.get(term) || 0) + 1);
    }
  }

  // Calculate IDF
  for (const [term, df] of docFrequency.entries()) {
    idf.set(term, Math.log(totalDocs / df));
  }

  return idf;
}

/**
 * Calculate TF-IDF vector for a document
 */
function calculateTFIDF(tf: Map<string, number>, idf: Map<string, number>): Map<string, number> {
  const tfidf = new Map<string, number>();

  for (const [term, termFreq] of tf.entries()) {
    const idfValue = idf.get(term) || 0;
    tfidf.set(term, termFreq * idfValue);
  }

  return tfidf;
}

/**
 * Calculate vector magnitude
 */
function vectorMagnitude(vector: Map<string, number>): number {
  let sumSquares = 0;
  for (const value of vector.values()) {
    sumSquares += value * value;
  }
  return Math.sqrt(sumSquares);
}

/**
 * Calculate cosine similarity between two vectors
 */
function cosineSimilarity(vec1: Map<string, number>, vec2: Map<string, number>): number {
  let dotProduct = 0;

  // Calculate dot product
  for (const [term, value1] of vec1.entries()) {
    const value2 = vec2.get(term);
    if (value2 !== undefined) {
      dotProduct += value1 * value2;
    }
  }

  const mag1 = vectorMagnitude(vec1);
  const mag2 = vectorMagnitude(vec2);

  if (mag1 === 0 || mag2 === 0) return 0;

  return dotProduct / (mag1 * mag2);
}

/**
 * Build TF-IDF vectors for all products
 */
export function buildProductVectors(products: Product[]): Map<string, DocumentVector> {
  // Prepare documents
  const documents = products.map(p => {
    const text = `${p.title} ${p.tags.join(' ')}`;
    return tokenize(text);
  });

  // Calculate IDF
  const idf = calculateIDF(documents);

  // Build vectors
  const vectors = new Map<string, DocumentVector>();

  for (let i = 0; i < products.length; i++) {
    const product = products[i];
    const tokens = documents[i];
    const tf = calculateTF(tokens);
    const tfidf = calculateTFIDF(tf, idf);

    vectors.set(product.id, {
      productId: product.id,
      vector: tfidf,
      magnitude: vectorMagnitude(tfidf),
    });
  }

  return vectors;
}

/**
 * Find similar products using cosine similarity
 */
export function findSimilarProducts(
  targetProductId: string,
  products: Product[],
  limit = 5
): Array<{ productId: string; similarity: number }> {
  const vectors = buildProductVectors(products);
  const targetVector = vectors.get(targetProductId);

  if (!targetVector) {
    return [];
  }

  const similarities: Array<{ productId: string; similarity: number }> = [];

  for (const [productId, vector] of vectors.entries()) {
    if (productId === targetProductId) continue;

    const similarity = cosineSimilarity(targetVector.vector, vector.vector);
    similarities.push({ productId, similarity });
  }

  // Sort by similarity and return top results
  return similarities
    .sort((a, b) => b.similarity - a.similarity)
    .slice(0, limit);
}

/**
 * Calculate similarity score for search query
 */
export function calculateQuerySimilarity(query: string, product: Product): number {
  const queryTokens = tokenize(query);
  const productText = `${product.title} ${product.tags.join(' ')}`;
  const productTokens = tokenize(productText);

  // Simple overlap coefficient
  const querySet = new Set(queryTokens);
  const productSet = new Set(productTokens);

  let matches = 0;
  for (const token of querySet) {
    if (productSet.has(token)) {
      matches++;
    }
  }

  return matches / Math.max(querySet.size, 1);
}

File: backend-go/internal/utils/similarity.go
package utils

import (
	"math"
	"regexp"
	"strings"

	"github.com/kimberlysue0003/ecommerce-mini/backend-go/internal/models"
)

// DocumentVector represents TF-IDF vector for a product
type DocumentVector struct {
	ProductID string
	Vector    map[string]float64
	Magnitude float64
}

// ProductSimilarity represents similarity score between products
type ProductSimilarity struct {
	ProductID  string  `json:"productId"`
	Similarity float64 `json:"similarity"`
}

// Tokenize text into words (lowercase, no punctuation)
func Tokenize(text string) []string {
	// Convert to lowercase
	text = strings.ToLower(text)

	// Remove punctuation
	reg := regexp.MustCompile(`[^\w\s]`)
	text = reg.ReplaceAllString(text, " ")

	// Split and filter
	tokens := strings.Fields(text)
	result := make([]string, 0, len(tokens))
	for _, token := range tokens {
		if len(token) > 0 {
			result = append(result, token)
		}
	}
	return result
}

// CalculateTF calculates term frequency for a document
func CalculateTF(tokens []string) map[string]float64 {
	tf := make(map[string]float64)
	totalTokens := float64(len(tokens))

	// Count occurrences
	for _, token := range tokens {
		tf[token]++
	}

	// Normalize by document length
	for term := range tf {
		tf[term] = tf[term] / totalTokens
	}

	return tf
}

// CalculateIDF calculates inverse document frequency
func CalculateIDF(documents [][]string) map[string]float64 {
	idf := make(map[string]float64)
	totalDocs := float64(len(documents))

	// Count documents containing each term
	docFrequency := make(map[string]int)
	for _, doc := range documents {
		uniqueTerms := make(map[string]bool)
		for _, term := range doc {
			uniqueTerms[term] = true
		}
		for term := range uniqueTerms {
			docFrequency[term]++
		}
	}

	// Calculate IDF
	for term, df := range docFrequency {
		idf[term] = math.Log(totalDocs / float64(df))
	}

	return idf
}

// CalculateTFIDF calculates TF-IDF vector for a document
func CalculateTFIDF(tf map[string]float64, idf map[string]float64) map[string]float64 {
	tfidf := make(map[string]float64)

	for term, termFreq := range tf {
		idfValue := idf[term]
		tfidf[term] = termFreq * idfValue
	}

	return tfidf
}

// VectorMagnitude calculates the magnitude of a vector
func VectorMagnitude(vector map[string]float64) float64 {
	sumSquares := 0.0
	for _, value := range vector {
		sumSquares += value * value
	}
	return math.Sqrt(sumSquares)
}

// CosineSimilarity calculates cosine similarity between two vectors
func CosineSimilarity(vec1, vec2 map[string]float64) float64 {
	dotProduct := 0.0

	// Calculate dot product
	for term, value1 := range vec1 {
		if value2, exists := vec2[term]; exists {
			dotProduct += value1 * value2
		}
	}

	mag1 := VectorMagnitude(vec1)
	mag2 := VectorMagnitude(vec2)

	if mag1 == 0 || mag2 == 0 {
		return 0
	}

	return dotProduct / (mag1 * mag2)
}

// BuildProductVectors builds TF-IDF vectors for all products
func BuildProductVectors(products []models.Product) map[string]DocumentVector {
	// Prepare documents
	documents := make([][]string, len(products))
	for i, p := range products {
		text := p.Title + " " + strings.Join(p.Tags, " ")
		documents[i] = Tokenize(text)
	}

	// Calculate IDF
	idf := CalculateIDF(documents)

	// Build vectors
	vectors := make(map[string]DocumentVector)

	for i, product := range products {
		tokens := documents[i]
		tf := CalculateTF(tokens)
		tfidf := CalculateTFIDF(tf, idf)

		vectors[product.ID] = DocumentVector{
			ProductID: product.ID,
			Vector:    tfidf,
			Magnitude: VectorMagnitude(tfidf),
		}
	}

	return vectors
}

// FindSimilarProducts finds similar products using cosine similarity
func FindSimilarProducts(targetProductID string, products []models.Product, limit int) []ProductSimilarity {
	vectors := BuildProductVectors(products)
	targetVector, exists := vectors[targetProductID]

	if !exists {
		return []ProductSimilarity{}
	}

	similarities := make([]ProductSimilarity, 0)

	for productID, vector := range vectors {
		if productID == targetProductID {
			continue
		}

		similarity := CosineSimilarity(targetVector.Vector, vector.Vector)
		similarities = append(similarities, ProductSimilarity{
			ProductID:  productID,
			Similarity: similarity,
		})
	}

	// Sort by similarity (descending)
	for i := 0; i < len(similarities)-1; i++ {
		for j := i + 1; j < len(similarities); j++ {
			if similarities[j].Similarity > similarities[i].Similarity {
				similarities[i], similarities[j] = similarities[j], similarities[i]
			}
		}
	}

	// Return top results
	if limit > len(similarities) {
		limit = len(similarities)
	}
	return similarities[:limit]
}

// CalculateQuerySimilarity calculates similarity score for search query
func CalculateQuerySimilarity(query string, product models.Product) float64 {
	queryTokens := Tokenize(query)
	productText := product.Title + " " + strings.Join(product.Tags, " ")
	productTokens := Tokenize(productText)

	// Simple overlap coefficient
	querySet := make(map[string]bool)
	for _, token := range queryTokens {
		querySet[token] = true
	}

	productSet := make(map[string]bool)
	for _, token := range productTokens {
		productSet[token] = true
	}

	matches := 0
	for token := range querySet {
		if productSet[token] {
			matches++
		}
	}

	if len(querySet) == 0 {
		return 0
	}

	return float64(matches) / float64(len(querySet))
}
